### DISCUSSION TITLE
How to parse dot's HTML-like strings?

### DISCUSSION THREAD

**Discussion by NeverGivinUp** [winnow-rs/winnow#423] [Q&A] [answered]
Parsing [dot-files](https://www.graphviz.org/doc/info/lang.html) is generally straight forward. I don't know how to implement HTML-like strings, though. An HTML-like string is separated by an opening and closing angular bracket `<...>`. An attribute looks something like this
```
label=<<a href="somewhere"><b>That</b> place</a>>
```
For a test from graphviz see [here](https://gitlab.com/graphviz/graphviz/-/blob/main/tests/1622_0.dot).

It seems, like a parser, that would like to recognize, would need to hold state, so as to be able to detect, if a matching number of opening and closing brackets have been found.  The repeating combinators, like `take_while`, only allow `Fn` tests (not `FnMut`), and `ContainsToken::contains_token` only allows `&self`, not `&mut self`.

Is there a way to parse HTML-Like strings?

**Comment by epage** [comment]
We will likely need a delimiter matching parser for #96 but not feeling ready to define that quite yet.

You could write this imperatively, something roughly like
- while unbalanced:
  - `(take_until(['<', '>']), any)``
  - Modify balance check according to the the `any` value (either `<` or `>`)
- Return `O=()`
- `.recognize()` the above

**Comment by NeverGivinUp** [comment]
Thanks @epage!

I imagine, the parsing would look like this
```
fn p_html_inner(input: &str) -> IResult<&str, ()> {
    let (mut input, bracket) = '<'.parse_peek(input)?;
    let mut balance = 1;
    while balance != 0 {
        let (remaining_input, (before_bracket, bracket)) =
            (take_until0(['<', '>']), any).parse_peek(input)?;
        match bracket {
            '<' => balance += 1,
            '>' => balance -= 1,
            _ => unreachable!("angular bracket is not a bracket"),
        }
        input = remaining_input;
    }
    Ok((input, ()))
}
```

However, I don't know what to call the `recognize` method on. Do I need to create a struct that implements `Parser` instead of writing a function?

**Reply by epage** [reply, marked as answer]
`unpeek(p_html_inner).recognize()`

`unpeek` takes the old nom-like purely functional function signature and turns it into an `impl Parser<I, O, E>`.

You could also do
```rust
fn p_html_inner(input: &mut &str) -> PResult<()> {
    let bracket = '<'.parse_next(input)?;
    let mut balance = 1;
    while balance != 0 {
        let (before_bracket, bracket) =
            (take_until0(['<', '>']), any).parse_next(input)?;
        match bracket {
            '<' => balance += 1,
            '>' => balance -= 1,
            _ => unreachable!("angular bracket is not a bracket"),
        }
    }
    Ok(())
}
```
and then you could do `p_html_inner.recognize()` since we have `impl Parser for FnMut...` for that signature.

**Reply by epage** [reply]
btw `cargo add winnow -F simd` to use simd to accelerate `take_until0`
